File: packages/agent/src/infrastructure/events/types/turn.ts
/**
 * @fileoverview Turn Events
 *
 * Events for turn-level failures and errors.
 */

import type { BaseEvent } from './base.js';

// =============================================================================
// Turn Events
// =============================================================================

/**
 * Turn failed event
 *
 * Emitted when a turn fails due to provider errors, context limits,
 * or other recoverable/non-recoverable errors. This event ensures
 * the iOS app receives visibility into failures.
 */
export interface TurnFailedEvent extends BaseEvent {
  type: 'turn.failed';
  payload: {
    /** Turn number that failed */
    turn: number;
    /** Human-readable error message */
    error: string;
    /** Error category code (e.g., 'PAUTH', 'PRATE', 'NET', 'CTX') */
    code?: string;
    /** Human-readable error category */
    category?: string;
    /** Whether the user can retry this operation */
    recoverable: boolean;
    /** Any content generated before the failure occurred */
    partialContent?: string;
  };
}

File: packages/agent/src/domains/session/event_store/types/payloads/turn.rs
//! Turn event payloads.

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Payload for `turn.failed` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnFailedPayload {
    /// Turn number.
    pub turn: i64,
    /// Error message.
    pub error: String,
    /// Error code (e.g., "PAUTH", "PRATE", "NET", "CTX").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
    /// Error category.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    /// Whether retrying the same request may succeed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub retryable: Option<bool>,
    /// Whether the error is recoverable.
    pub recoverable: bool,
    /// Layer that classified the failure.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub origin: Option<String>,
    /// Structured failure details. New runtime rows include
    /// `details.failure`, the canonical failure envelope.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<Value>,
    /// Content generated before failure.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub partial_content: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn turn_failed_payload_preserves_canonical_failure_details() {
        let raw = json!({
            "turn": 3,
            "error": "Interrupted by user",
            "code": "RUNTIME_CANCELLED",
            "category": "cancelled",
            "retryable": false,
            "recoverable": true,
            "origin": "agent_runtime",
            "details": {
                "failure": {
                    "code": "RUNTIME_CANCELLED",
                    "category": "cancelled",
                    "message": "Interrupted by user",
                    "retryable": false,
                    "recoverable": true,
                    "origin": "agent_runtime"
                }
            },
            "partialContent": null
        });

        let payload: TurnFailedPayload = serde_json::from_value(raw).unwrap();

        assert_eq!(payload.code.as_deref(), Some("RUNTIME_CANCELLED"));
        assert_eq!(payload.category.as_deref(), Some("cancelled"));
        assert_eq!(payload.retryable, Some(false));
        assert_eq!(payload.origin.as_deref(), Some("agent_runtime"));
        assert_eq!(
            payload.details.unwrap()["failure"]["code"],
            "RUNTIME_CANCELLED"
        );
    }
}
